cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt

```diff
@@ -0,0 +1,305 @@
+        return loadersSynchronizer.synchronizedFor(key) {
+            val nowNanos = clock.currentTimeNanos
+            cacheEntries[key]?.let {
+                if (it.isExpired(nowNanos)) {
+                    // clean up expired entries
+                    expireEntries(nowNanos)
+                    null
+                } else {
+                    // update eviction metadata
+                    recordRead(it, nowNanos)
+                    it.value
+                }
+            } ?: loader().let { loadedValue ->
+                // cache the loaded value if value for the key is still absent
+                // from the cache entries after executing the loader
+                val existingValue = get(key)
+                if (existingValue != null) {
+                    existingValue
+                } else {
+                    put(key, loadedValue)
+                    loadedValue
+                }
+            }
+        }
+    }
+
+    override fun put(key: Key, value: Value) {
+        val nowNanos = clock.currentTimeNanos
+        expireEntries(nowNanos)
+
+        val existingEntry = cacheEntries[key]
+        if (existingEntry != null) {
+            // cache entry found
+            recordWrite(existingEntry, nowNanos)
+            existingEntry.value = value
+        } else {
+            // create a new cache entry
+            val newEntry = CacheEntry(key, value)
+            recordWrite(newEntry, nowNanos)
+            cacheEntries[key] = newEntry
+        }
+
+        evictEntries()
+    }
+
+    override fun invalidate(key: Key) {
+        val nowNanos = clock.currentTimeNanos
+        expireEntries(nowNanos)
+
+        cacheEntries.remove(key)?.also {
+            writeQueue?.remove(it)
+            accessQueue?.remove(it)
+        }
+    }
+
+    override fun invalidateAll() {
+        cacheEntries.clear()
+        writeQueue?.clear()
+        accessQueue?.clear()
+    }
+
+    /**
+     * Remove all expired entries.
+     */
+    private fun expireEntries(nowNanos: Long) {
+        val queuesToProcess = listOfNotNull(
+            if (expiresAfterWrite) writeQueue else null,
+            if (expiresAfterAccess) accessQueue else null
+        )
+
+        // address any inconsistencies between the queues and map before processing.
+        if (queuesToProcess.isNotEmpty()) {
+            cleanUpDeadEntries()
+        }
+
+        queuesToProcess.forEach { queue ->
+            synchronized(queue) {
+                val iterator = queue.iterator()
+                for (entry in iterator) {
+                    if (entry.isExpired(nowNanos)) {
+                        cacheEntries.remove(entry.key)
+                        // remove the entry from the current queue
+                        iterator.remove()
+                        // also remove the entry from the other queue
+                        if (queue == writeQueue) {
+                            accessQueue?.remove(entry)
```

Comment:
There's a hidden deadlock here. Sync set uses the set as the lock for remove. If 2 threads are evicting entries they can deadlock on each other. Remove this optimization. You will remove the entry when iterating on the next queue

Reply:
I mean remove the entire if-else removes

Reply:
Good catch. Thanks.